expr_checker.py

```diff
@@ -0,0 +1,114 @@
+
+import ast
+from copy import copy
+
+
+_safe_type = None
+_readonly = None
+
+
+def check_type(obj):
+    if type(obj) in {list, tuple, set}:
+        list(map(check_type, obj))
+        return obj
+    elif type(obj) not in _safe_type:
+        raise ValueError(f"safe_eval didn't like {obj}")
+    else:
+        return obj
+
+
+def test(func, *args, **kwargs):
+    list(map(check_type, kwargs.values()))
+    list(map(check_type, args))
+
+    return check_type(func(*args, **kwargs))
+
+
+def check_ro(obj):
+    if type(obj) in _readonly:
+        return copy(obj)
+    else:
+        return obj
+
+
+class NodeChecker(ast.NodeTransformer):
+    def __init__(self, check_fn, allowed_attr):
+        self.check_fn = check_fn.__name__
+        self.allowed_attr = allowed_attr
+
+        super().__init__()
+
+    def visit_Call(self, node):
+        list(map(self.visit, node.args))
+
+        if isinstance(node.func, ast.Attribute):
+            if node.func.attr not in self.allowed_attr:
+                raise ValueError(
+                    f"safe_eval doesn't allow you to read {node.func.attr}")
+
+            name = copy(node.func.value)
+            node.func.value.__class__ = ast.Call
+            node.func.value.func = ast.Name("check_ro", ctx=ast.Load())
+            node.func.value.args = [name]
+            node.func.value.keywords = []
+        
+        else:
+            self.generic_visit(node.func)            
+            
+        return ast.Call(
+            func=ast.Name(self.check_fn, ctx=ast.Load()),
+            args=[node.func] + node.args,
+            keywords=node.keywords
+        )
+
+    def visit_FunctionDef(self, node):
+        if node.name in [self.check_fn, "check_ro", "check_type"]:
+            raise NameError(f"safe_eval: {node.name} is a reserved name")
+        
+        return node
+
+    def visit_Name(self, node):
+        if node.id in [self.check_fn, "check_ro", "check_type"]:
+            raise NameError(f"safe_eval: {node.id} is a reserved name")
+
+        return node
+
+    def visit_Attribute(self, node):
+        if node.attr not in self.allowed_attr:
+            raise ValueError(
+                f"safe_eval doesn't allow you to read {node.attr}")
+
+        if isinstance(node.ctx, ast.Load):
+            return ast.Call(
+                func=ast.Name("check_type", ctx=ast.Load()),
+                args=[node],
+                keywords=[]
+            )
+
+        else:
+            return ast.Call(
+                func=ast.Name("check_ro", ctx=ast.Load()),
+                args=[node.value],
+                keywords=[]
+            )
+
+    def visit_Assign(self, node):
+        list(map(self.visit, node.targets))
+        return ast.Call(
+            func=ast.Name("check_type", ctx=ast.Load()),
+            args=[node.value],
+            keywords=[]
+        )
+
+
+def expr_checker(expr, whitelist, allowed_attr, readonly, check_function=test):
+    global _safe_type
+    global _readonly
```

Comment:
Do you see any alternatives ?

I have tried to pass it to `NodeChecker` as an attribute and then give it inside the ast but it won't work :sweat_smile:

Reply:
Exactly like you do for `check_fn` (give to NodeChecker and store then in the object attributes)

Why won't that work ?

Reply:
Ast works mostly on strings. with `check_fn` I pass the function name and voilà it works. 

But in _readonly I have a set of class names that looks a bit like that: `{<class '__main__.ReadOnlyObject'>}`. 
When you give that type of strings back to the Ast, it won't like it at all (because it's not valide Python code)

Reply:
Anyway, if `_readonly` is in the globals, `check_ro` won't work when the globals will be cleaned... (or when you copy / paste the generated code in a python console)

But I have an other comment to replace this list by a function `__ast_check_get_attr`, we can imagine that `__ast_check_get_attr` return the value, and so we can make the copy there (we will have less argument, and better control on what is done)

So `_readonly` must be replaced by `__ast_check_get_attr` (and adapt the related test) and this method should be added in the generated code like you do with `check_type` and `check_function`